rfd/0164-scoped-rbac.md

```diff
@@ -0,0 +1,956 @@
+
+## User stories
+
+Let’s get back to the issues we outlined  in the start of this RFD and review how the new system will  help to resolve them.
+
+### Gradual scoping
+
+Most Teleport customers will start with all resources and roles in the cluster scope. 
+
+We will let them introduce resource groups gradually. Let’s create two resource groups, `prod` and `dev` with resource groups `west` and `lab`.
+
+Here is the resource groups hierarchy, where we will assume that mars and luna servers matched the assignments:
+
+```mermaid
+graph TD;
+    luna(Server luna)-->west;
+    west-->prod;
+    prod-->cluster;
+    lab-->dev;
+    mars(Server mars)-->lab;
+    dev-->cluster;
+```
+
+```yaml
+kind: resource_group
+metadata:
+  name: prod
+---
+kind: resource_group
+metadata:
+  name: west
+  parent_resource_group: prod
+  # would be nice if we could match on AWS specific right away with match_aws
+  match_aws:
+    account_id: aws-account-id
+    region: west
+---
+kind: resource_group
+metadata:
+  name: dev
+---
+kind: resource_group
+metadata:
+  name: lab
+spec:
+  parent_resource_group: dev
+  # here we will just match on labels
+  match_labels:
+    kinds: [node]
+    env: lab
+```
+
+This will let administrators create a resource hierarchy by  mapping computing resources using AWS metadata or labels.
+
+We will use this setup in the following examples.
+
+### SSH access to specific hosts
+
+The most prominent use-case is our over-engineered access role. We can keep this role as is. Today, it grants blanket access to any computing resource of Teleport.
+
+Alice, who is an administrator, would like to restrict access for a user bob@example.com to any server in the lab as root
+
+```yaml
+kind: access_list
+metadata:
+  name: access-to-lab
+spec:
+  grants: 
+    roles: [access]
+    traits:
+      'internal.logins' : 'root'
+  scope: '/dev/lab`
+  members:
+    - bob@example.com
+```
+
+Teleport will grant role access  and traits internal.logins: root to `bob@example.com`, but only when Bob would access servers in the resource group `/dev/lab`. 
+
+This grant will not be valid out of the scope of `/dev/lab`, so Bob won’t be able to SSH as root to any other servers.
+
+K8s access to specific clusters
+
+Teleport can autodiscover clusters echo and bravo with namespaces default and prod, creating the following resource group hierarchy:
+
+```
+/k8s/namespaces/prod/bravo
+/k8s/namespaces/prod/echo
+
+/k8s/namespaces/default/bravo
+/k8s/namespaces/default/echo
+```
+
+Note that here we have set namespaces, and not cluster names as the root of the resource hierarchy,
+so we can group different cluster names by namespace.
+
+We can then use this hierarchy to create access lists specifying access to default namespace in any cluster:
+
+```yaml
+kind: access_list
+metadata:
+  name: access-to-default
+spec:
+  grants: 
+    roles: [access]
+    traits:
+      'internal.logins' : 'root'
+  scope:  '/k8s/namespaces/default`
```

Comment:
Azure has a nice feature that decouples the resource you access from the data you are allowed to see in the resource.

they differentiate it between `Actions` and `DataActions` [check here](https://goteleport.com/docs/kubernetes-access/discovery/azure/)
If we can introduce the same idea, we could generalize this a bit more like the following:

```yaml
kind: access_list
metadata:
  name: access-to-default
spec:
  grants: 
    roles: [access]
    traits:
      'internal.logins' : 'root'
  scope:  '/dev/lab'
  dataScopes: [ '/k8s/namespaces/default', '/k8s/namespaces/kube-system/pods/kube-dns']
  members:
    - bob@example.com
```

With this, you clearly see that the access list grants access to resources members or indirect members of `resource_group` `/dev/lab` but members of the access list will have a data scope reduced to `namespace=default` and to `kube-dns` pod from `kube-system` namespace.
The same applies to database access where the dataScope will be `table, database`...

With this approach, Teleport doesn't need to synchronize the K8S cluster scopes given that everything can be evaluated at runtime.

Reply:
traits should be OK for this purpose, we should keep using them instead. They provide the same features as data scopes

Reply:
I might have missed something.

So what you propose is to leverage the role with the template expansion?

So a role with the following `kubernetes_resources`,

```yaml
    kubernetes_resources:
    - kind: namespace
      name: '{{traits.namespaces}}'
      verbs:
      - '*'

```

will expand to something like the snippet below for a user with trait namespace = default, kube-system.

```yaml
    kubernetes_resources:
    - kind: namespace
      name: [default,kube-system]
      verbs:
      - '*'
```

Is this what you meant or will we magically map random traits into some restrictions?

Reply:
> So what you propose is to leverage the role with the template expansion?

yes